sources/work/busybox-1.36.1/docs/busybox.1
.\" -*- mode: troff; coding: utf-8 -*-
.\" Automatically generated by Pod::Man 5.01 (Pod::Simple 3.43)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" \*(C` and \*(C' are quotes in nroff, nothing in troff, for use with C<>.
.ie n \{\
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\" ========================================================================
.\"
.IX Title "BUSYBOX 1"
.TH BUSYBOX 1 2025-10-25 "version 1.36.1" busybox
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH NAME
BusyBox \- The Swiss Army Knife of Embedded Linux
.SH SYNTAX
.IX Header "SYNTAX"
.Vb 1
\& busybox <applet> [arguments...]  # or
\&
\& <applet> [arguments...]          # if symlinked
.Ve
.SH DESCRIPTION
.IX Header "DESCRIPTION"
BusyBox combines tiny versions of many common UNIX utilities into a single
small executable. It provides minimalist replacements for most of the utilities
you usually find in GNU coreutils, util-linux, etc. The utilities in BusyBox
generally have fewer options than their full-featured GNU cousins; however, the
options that are included provide the expected functionality and behave very
much like their GNU counterparts.
.PP
BusyBox has been written with size-optimization and limited resources in mind.
It is also extremely modular so you can easily include or exclude commands (or
features) at compile time. This makes it easy to customize your embedded
systems. To create a working system, just add /dev, /etc, and a Linux kernel.
BusyBox provides a fairly complete POSIX environment for any small or embedded
system.
.PP
BusyBox is extremely configurable.  This allows you to include only the
components you need, thereby reducing binary size. Run 'make config' or 'make
menuconfig' to select the functionality that you wish to enable.  Then run
\&'make' to compile BusyBox using your configuration.
.PP
After the compile has finished, you should use 'make install' to install
BusyBox. This will install the 'bin/busybox' binary, in the target directory
specified by CONFIG_PREFIX. CONFIG_PREFIX can be set when configuring BusyBox,
or you can specify an alternative location at install time (i.e., with a
command line like 'make CONFIG_PREFIX=/tmp/foo install'). If you enabled
any applet installation scheme (either as symlinks or hardlinks), these will
also be installed in the location pointed to by CONFIG_PREFIX.
.SH USAGE
.IX Header "USAGE"
BusyBox is a multi-call binary.  A multi-call binary is an executable program
that performs the same job as more than one utility program.  That means there
is just a single BusyBox binary, but that single binary acts like a large
number of utilities.  This allows BusyBox to be smaller since all the built-in
utility programs (we call them applets) can share code for many common
operations.
.PP
You can also invoke BusyBox by issuing a command as an argument on the
command line.  For example, entering
.PP
.Vb 1
\&        /bin/busybox ls
.Ve
.PP
will also cause BusyBox to behave as 'ls'.
.PP
Of course, adding '/bin/busybox' into every command would be painful.  So most
people will invoke BusyBox using links to the BusyBox binary.
.PP
For example, entering
.PP
.Vb 2
\&        ln \-s /bin/busybox ls
\&        ./ls
.Ve
.PP
will cause BusyBox to behave as 'ls' (if the 'ls' command has been compiled
into BusyBox).  Generally speaking, you should never need to make all these
links yourself, as the BusyBox build system will do this for you when you run
the 'make install' command.
.PP
If you invoke BusyBox with no arguments, it will provide you with a list of the
applets that have been compiled into your BusyBox binary.
.SH "COMMON OPTIONS"
.IX Header "COMMON OPTIONS"
Most BusyBox applets support the \fB\-\-help\fR argument to provide a terse runtime
description of their behavior.  If the CONFIG_FEATURE_VERBOSE_USAGE option has
been enabled, more detailed usage information will also be available.
.SH COMMANDS
.IX Header "COMMANDS"
Currently available applets include:
.PP
.Vb 10
\&        [, [[, acpid, add\-shell, addgroup, adduser, adjtimex, arch, arp,
\&        arping, ascii, ash, awk, base32, base64, basename, bc, beep,
\&        blkdiscard, blkid, blockdev, bootchartd, brctl, bunzip2, bzcat,
\&        bzip2, cal, cat, chat, chattr, chgrp, chmod, chown, chpasswd, chpst,
\&        chroot, chrt, chvt, cksum, clear, cmp, comm, conspy, cp, cpio,
\&        crc32, crond, crontab, cryptpw, cttyhack, cut, date, dc, dd,
\&        deallocvt, delgroup, deluser, depmod, devmem, df, dhcprelay, diff,
\&        dirname, dmesg, dnsd, dnsdomainname, dos2unix, dpkg, dpkg\-deb, du,
\&        dumpkmap, dumpleases, echo, ed, egrep, eject, env, envdir,
\&        envuidgid, ether\-wake, expand, expr, factor, fakeidentd, fallocate,
\&        false, fatattr, fbset, fbsplash, fdflush, fdformat, fdisk,
\&        fgconsole, fgrep, find, findfs, flock, fold, free, freeramdisk,
\&        fsck, fsck.minix, fsfreeze, fstrim, fsync, ftpd, ftpget, ftpput,
\&        fuser, getopt, getty, grep, groups, gunzip, gzip, halt, hd, hdparm,
\&        head, hexdump, hexedit, hostid, hostname, httpd, hush, hwclock,
\&        i2cdetect, i2cdump, i2cget, i2cset, i2ctransfer, id, ifconfig,
\&        ifdown, ifenslave, ifplugd, ifup, inetd, init, insmod, install,
\&        ionice, iostat, ip, ipaddr, ipcalc, ipcrm, ipcs, iplink, ipneigh,
\&        iproute, iprule, iptunnel, kbd_mode, kill, killall, killall5, klogd,
\&        last, less, link, linux32, linux64, linuxrc, ln, loadfont, loadkmap,
\&        logger, login, logname, logread, losetup, lpd, lpq, lpr, ls, lsattr,
\&        lsmod, lsof, lspci, lsscsi, lsusb, lzcat, lzma, lzop, makedevs,
\&        makemime, man, md5sum, mdev, mesg, microcom, mim, mkdir, mkdosfs,
\&        mke2fs, mkfifo, mkfs.ext2, mkfs.minix, mkfs.vfat, mknod, mkpasswd,
\&        mkswap, mktemp, modinfo, modprobe, more, mount, mountpoint, mpstat,
\&        mt, mv, nameif, nanddump, nandwrite, nbd\-client, nc, netstat, nice,
\&        nl, nmeter, nohup, nologin, nproc, nsenter, nslookup, ntpd, od,
\&        openvt, partprobe, passwd, paste, patch, pgrep, pidof, ping, ping6,
\&        pipe_progress, pivot_root, pkill, pmap, popmaildir, poweroff,
\&        powertop, printenv, printf, ps, pscan, pstree, pwd, pwdx,
\&        raidautorun, rdate, rdev, readahead, readlink, readprofile,
\&        realpath, reboot, reformime, remove\-shell, renice, reset, resize,
\&        resume, rev, rm, rmdir, rmmod, route, rpm, rpm2cpio, rtcwake,
\&        run\-init, run\-parts, runlevel, runsv, runsvdir, rx, script,
\&        scriptreplay, sed, seedrng, sendmail, seq, setarch, setconsole,
\&        setfattr, setfont, setkeycodes, setlogcons, setpriv, setserial,
\&        setsid, setuidgid, sh, sha1sum, sha256sum, sha3sum, sha512sum,
\&        showkey, shred, shuf, slattach, sleep, smemcap, softlimit, sort,
\&        split, ssl_client, start\-stop\-daemon, stat, strings, stty, su,
\&        sulogin, sum, sv, svc, svlogd, svok, swapoff, swapon, switch_root,
\&        sync, sysctl, syslogd, tac, tail, tar, taskset, tcpsvd, tee, telnet,
\&        telnetd, test, tftp, tftpd, time, timeout, top, touch, tr,
\&        traceroute, traceroute6, tree, true, truncate, ts, tsort, tty,
\&        ttysize, tunctl, ubiattach, ubidetach, ubimkvol, ubirename,
\&        ubirmvol, ubirsvol, ubiupdatevol, udhcpc, udhcpc6, udhcpd, udpsvd,
\&        uevent, umount, uname, unexpand, uniq, unix2dos, unlink, unlzma,
\&        unshare, unxz, unzip, uptime, users, usleep, uudecode, uuencode,
\&        vconfig, vi, vlock, volname, w, wall, watch, watchdog, wc, wget,
\&        which, who, whoami, whois, xargs, xxd, xz, xzcat, yes, zcat, zcip
.Ve
.SH "COMMAND DESCRIPTIONS"
.IX Header "COMMAND DESCRIPTIONS"
.IP \fBacpid\fR 4
.IX Item "acpid"
acpid [\-df] [\-c CONFDIR] [\-l LOGFILE] [\-a ACTIONFILE] [\-M MAPFILE] [\-e PROC_EVENT_FILE] [\-p PIDFILE]
.Sp
Listen to ACPI events and spawn specific helpers on event arrival
.Sp
.Vb 8
\&        \-d      Log to stderr, not log file (implies \-f)
\&        \-f      Run in foreground
\&        \-c DIR  Config directory [/etc/acpi]
\&        \-e FILE /proc event file [/proc/acpi/event]
\&        \-l FILE Log file [/var/log/acpid.log]
\&        \-p FILE Pid file [/var/run/acpid.pid]
\&        \-a FILE Action file [/etc/acpid.conf]
\&        \-M FILE Map file [/etc/acpi.map]
.Ve
.Sp
Accept and ignore compatibility options \-g \-m \-s \-S \-v
.IP \fBadd-shell\fR 4
.IX Item "add-shell"
add-shell SHELL...
.Sp
Add SHELLs to /etc/shells
.IP \fBaddgroup\fR 4
.IX Item "addgroup"
addgroup [\-g GID] [\-S] [USER] GROUP
.Sp
Add a group or add a user to a group
.Sp
.Vb 2
\&        \-g GID  Group id
\&        \-S      Create a system group
.Ve
.IP \fBadduser\fR 4
.IX Item "adduser"
adduser [OPTIONS] USER [GROUP]
.Sp
Create new user, or add USER to GROUP
.Sp
.Vb 9
\&        \-h DIR          Home directory
\&        \-g GECOS        GECOS field
\&        \-s SHELL        Login shell
\&        \-G GRP          Group
\&        \-S              Create a system user
\&        \-D              Don\*(Aqt assign a password
\&        \-H              Don\*(Aqt create home directory
\&        \-u UID          User id
\&        \-k SKEL         Skeleton directory (/etc/skel)
.Ve
.IP \fBadjtimex\fR 4
.IX Item "adjtimex"
adjtimex [\-q] [\-o OFS] [\-f FREQ] [\-p TCONST] [\-t TICK]
.Sp
Read or set kernel time variables. See \fBadjtimex\fR\|(2)
.Sp
.Vb 6
\&        \-q      Quiet
\&        \-o OFF  Time offset, microseconds
\&        \-f FREQ Frequency adjust, integer kernel units (65536 is 1ppm)
\&        \-t TICK Microseconds per tick, usually 10000
\&                (positive \-t or \-f values make clock run faster)
\&        \-p TCONST
.Ve
.IP \fBarch\fR 4
.IX Item "arch"
arch
.Sp
Print system architecture
.IP \fBarp\fR 4
.IX Item "arp"
arp 
[\-vn]	[\-H HWTYPE] [\-i IF] \-a [HOSTNAME]
[\-v]		    [\-i IF] \-d HOSTNAME [pub]
[\-v]	[\-H HWTYPE] [\-i IF] \-s HOSTNAME HWADDR [temp]
[\-v]	[\-H HWTYPE] [\-i IF] \-s HOSTNAME HWADDR [netmask MASK] pub
[\-v]	[\-H HWTYPE] [\-i IF] \-Ds HOSTNAME IFACE [netmask MASK] pub
.Sp
Manipulate ARP cache
.Sp
.Vb 9
\&        \-a              Display (all) hosts
\&        \-d              Delete ARP entry
\&        \-s              Set new entry
\&        \-v              Verbose
\&        \-n              Don\*(Aqt resolve names
\&        \-i IF           Network interface
\&        \-D              Read HWADDR from IFACE
\&        \-A,\-p AF        Protocol family
\&        \-H HWTYPE       Hardware address type
.Ve
.IP \fBarping\fR 4
.IX Item "arping"
arping [\-fqbDUA] [\-c CNT] [\-w TIMEOUT] [\-I IFACE] [\-s SRC_IP] DST_IP
.Sp
Send ARP requests/replies
.Sp
.Vb 11
\&        \-f              Quit on first ARP reply
\&        \-q              Quiet
\&        \-b              Keep broadcasting, don\*(Aqt go unicast
\&        \-D              Exit with 1 if DST_IP replies
\&        \-U              Unsolicited ARP mode, update your neighbors
\&        \-A              ARP answer mode, update your neighbors
\&        \-c N            Stop after sending N ARP requests
\&        \-w TIMEOUT      Seconds to wait for ARP reply
\&        \-I IFACE        Interface to use (default eth0)
\&        \-s SRC_IP       Sender IP address
\&        DST_IP          Target IP address
.Ve
.IP \fBash\fR 4
.IX Item "ash"
ash [\-il] [\-|+Cabefmnuvx] [\-|+o OPT]... [\-c 'SCRIPT' [ARG0 ARGS] | FILE ARGS | \-s ARGS]
.Sp
Unix shell interpreter
.IP \fBawk\fR 4
.IX Item "awk"
awk [OPTIONS] [AWK_PROGRAM] [FILE]...
.Sp
.Vb 4
\&        \-v VAR=VAL      Set variable
\&        \-F SEP          Use SEP as field separator
\&        \-f FILE         Read program from FILE
\&        \-e AWK_PROGRAM
.Ve
.IP \fBbase32\fR 4
.IX Item "base32"
base32 [\-d] [\-w COL] [FILE]
.Sp
Base32 encode or decode FILE to standard output
.Sp
.Vb 2
\&        \-d      Decode data
\&        \-w COL  Wrap lines at COL (default 76, 0 disables)
.Ve
.IP \fBbase64\fR 4
.IX Item "base64"
base64 [\-d] [\-w COL] [FILE]
.Sp
Base64 encode or decode FILE to standard output
.Sp
.Vb 2
\&        \-d      Decode data
\&        \-w COL  Wrap lines at COL (default 76, 0 disables)
.Ve
.IP \fBbasename\fR 4
.IX Item "basename"
basename FILE [SUFFIX] | \-a FILE... | \-s SUFFIX FILE...
.Sp
Strip directory path and SUFFIX from FILE
.Sp
.Vb 2
\&        \-a              All arguments are FILEs
\&        \-s SUFFIX       Remove SUFFIX (implies \-a)
.Ve
.IP \fBbc\fR 4
.IX Item "bc"
bc [\-sqlw] [FILE]...
.Sp
Arbitrary precision calculator
.Sp
.Vb 4
\&        \-q      Quiet
\&        \-l      Load standard library
\&        \-s      Be POSIX compatible
\&        \-w      Warn if extensions are used
.Ve
.Sp
\&\f(CW$BC_LINE_LENGTH\fR changes output width
.IP \fBbeep\fR 4
.IX Item "beep"
beep \-f FREQ \-l LEN \-d DELAY \-r COUNT \-n
.Sp
.Vb 5
\&        \-f      Frequency in Hz
\&        \-l      Length in ms
\&        \-d      Delay in ms
\&        \-r      Repetitions
\&        \-n      Start new tone
.Ve
.IP \fBblkdiscard\fR 4
.IX Item "blkdiscard"
blkdiscard [\-o OFS] [\-l LEN] [\-s] DEVICE
.Sp
Discard sectors on DEVICE
.Sp
.Vb 3
\&        \-o OFS  Byte offset into device
\&        \-l LEN  Number of bytes to discard
\&        \-s      Perform a secure discard
.Ve
.IP \fBblkid\fR 4
.IX Item "blkid"
blkid [BLOCKDEV]...
.Sp
Print UUIDs of all filesystems
.IP \fBblockdev\fR 4
.IX Item "blockdev"
blockdev OPTION BLOCKDEV
.Sp
.Vb 12
\&        \-\-setro         Set ro
\&        \-\-setrw         Set rw
\&        \-\-getro         Get ro
\&        \-\-getss         Get sector size
\&        \-\-getbsz        Get block size
\&        \-\-setbsz BYTES  Set block size
\&        \-\-getsz         Get device size in 512\-byte sectors
\&        \-\-getsize64     Get device size in bytes
\&        \-\-getra         Get readahead in 512\-byte sectors
\&        \-\-setra SECTORS Set readahead
\&        \-\-flushbufs     Flush buffers
\&        \-\-rereadpt      Reread partition table
.Ve
.IP \fBbootchartd\fR 4
.IX Item "bootchartd"
bootchartd start [PROG ARGS]|stop|init
.Sp
Create /var/log/bootchart.tgz with boot chart data
.Sp
start: start background logging; with PROG, run PROG, then kill logging with USR1
stop: send USR1 to all bootchartd processes
init: start background logging; stop when getty/xdm is seen (for init scripts)
Under PID 1: as init, then exec \f(CW$bootchart_init\fR, /init, /sbin/init
.IP \fBbrctl\fR 4
.IX Item "brctl"
brctl COMMAND [BRIDGE [ARGS]]
.Sp
Manage ethernet bridges
Commands:
.Sp
.Vb 10
\&        show [BRIDGE]...        Show bridges
\&        addbr BRIDGE            Create BRIDGE
\&        delbr BRIDGE            Delete BRIDGE
\&        addif BRIDGE IFACE      Add IFACE to BRIDGE
\&        delif BRIDGE IFACE      Delete IFACE from BRIDGE
\&        showmacs BRIDGE                 List MAC addresses
\&        showstp BRIDGE                  Show STP info
\&        stp BRIDGE 1/yes/on|0/no/off    Set STP on/off
\&        setageing BRIDGE SECONDS        Set ageing time
\&        setfd BRIDGE SECONDS            Set bridge forward delay
\&        sethello BRIDGE SECONDS         Set hello time
\&        setmaxage BRIDGE SECONDS        Set max message age
\&        setbridgeprio BRIDGE PRIO       Set bridge priority
\&        setportprio BRIDGE IFACE PRIO   Set port priority
\&        setpathcost BRIDGE IFACE COST   Set path cost
.Ve
.IP \fBbunzip2\fR 4
.IX Item "bunzip2"
bunzip2 [\-cfk] [FILE]...
.Sp
Decompress FILEs (or stdin)
.Sp
.Vb 4
\&        \-c      Write to stdout
\&        \-f      Force
\&        \-k      Keep input files
\&        \-t      Test integrity
.Ve
.IP \fBbzcat\fR 4
.IX Item "bzcat"
bzcat [FILE]...
.Sp
Decompress to stdout
.IP \fBbzip2\fR 4
.IX Item "bzip2"
bzip2 [\-cfkdt123456789] [FILE]...
.Sp
Compress FILEs (or stdin) with bzip2 algorithm
.Sp
.Vb 6
\&        \-1..9   Compression level
\&        \-d      Decompress
\&        \-c      Write to stdout
\&        \-f      Force
\&        \-k      Keep input files
\&        \-t      Test integrity
.Ve
.IP \fBcal\fR 4
.IX Item "cal"
cal [\-jmy] [[MONTH] YEAR]
.Sp
Display a calendar
.Sp
.Vb 3
\&        \-j      Use julian dates
\&        \-m      Week starts on Monday
\&        \-y      Display the entire year
.Ve
.IP \fBcat\fR 4
.IX Item "cat"
cat [\-nbvteA] [FILE]...
.Sp
Print FILEs to stdout
.Sp
.Vb 6
\&        \-n      Number output lines
\&        \-b      Number nonempty lines
\&        \-v      Show nonprinting characters as ^x or M\-x
\&        \-t      ...and tabs as ^I
\&        \-e      ...and end lines with $
\&        \-A      Same as \-vte
.Ve
.IP \fBchat\fR 4
.IX Item "chat"
chat EXPECT [SEND [EXPECT [SEND]]...]
.Sp
Useful for interacting with a modem connected to stdin/stdout.
A script consists of "expect-send" argument pairs.
Example:
.Sp
chat '' ATZ OK ATD123456 CONNECT '' ogin: pppuser word: ppppass '~'
.IP \fBchattr\fR 4
.IX Item "chattr"
chattr [\-R] [\-v VERSION] [\-p PROJID] [\-+=AacDdijsStTu] FILE...
.Sp
Change ext2 file attributes
.Sp
.Vb 4
\&        \-R      Recurse
\&        \-v NUM  Set version/generation number
\&        \-p NUM  Set project number
\&Modifiers:
\&
\&        \-,+,=   Remove/add/set attributes
\&Attributes:
\&
\&        A       No atime
\&        a       Append only
\&        C       No copy\-on\-write
\&        c       Compressed
\&        D       Synchronous dir updates
\&        d       Don\*(Aqt backup with dump
\&        E       Encrypted
\&        e       File uses extents
\&        F       Case\-insensitive dir
\&        I       Indexed dir
\&        i       Immutable
\&        j       Write data to journal first
\&        N       File is stored in inode
\&        P       Hierarchical project ID dir
\&        S       Synchronous file updates
\&        s       Zero storage when deleted
\&        T       Top of dir hierarchy
\&        t       Don\*(Aqt tail\-merge with other files
\&        u       Allow undelete
\&        V       Verity
.Ve
.IP \fBchgrp\fR 4
.IX Item "chgrp"
chgrp [\-RhLHPcvf]... GROUP FILE...
.Sp
Change the group membership of FILEs to GROUP
.Sp
.Vb 8
\&        \-h      Affect symlinks instead of symlink targets
\&        \-L      Traverse all symlinks to directories
\&        \-H      Traverse symlinks on command line only
\&        \-P      Don\*(Aqt traverse symlinks (default)
\&        \-R      Recurse
\&        \-c      List changed files
\&        \-v      Verbose
\&        \-f      Hide errors
.Ve
.IP \fBchmod\fR 4
.IX Item "chmod"
chmod [\-Rcvf] MODE[,MODE]... FILE...
.Sp
MODE is octal number (bit pattern sstrwxrwxrwx) or [ugoa]{+|\-|=}[rwxXst]
.Sp
.Vb 4
\&        \-R      Recurse
\&        \-c      List changed files
\&        \-v      Verbose
\&        \-f      Hide errors
.Ve
.IP \fBchown\fR 4
.IX Item "chown"
chown [\-RhLHPcvf]... USER[:[GRP]] FILE...
.Sp
Change the owner and/or group of FILEs to USER and/or GRP
.Sp
.Vb 8
\&        \-h      Affect symlinks instead of symlink targets
\&        \-L      Traverse all symlinks to directories
\&        \-H      Traverse symlinks on command line only
\&        \-P      Don\*(Aqt traverse symlinks (default)
\&        \-R      Recurse
\&        \-c      List changed files
\&        \-v      Verbose
\&        \-f      Hide errors
.Ve
.IP \fBchpasswd\fR 4
.IX Item "chpasswd"
chpasswd [\-me] [\-c ALG] [\-R DIR]
.Sp
Read user:password from stdin and update /etc/passwd
.Sp
.Vb 4
\&        \-e      Supplied passwords are in encrypted form
\&        \-m      Encrypt using md5, not des
\&        \-c ALG  des,md5,sha256/512 (default des)
\&        \-R DIR  Directory to chroot into
.Ve
.IP \fBchpst\fR 4
.IX Item "chpst"
chpst [\-vP012] [\-u USER[:GRP]] [\-U USER[:GRP]] [\-e DIR]
	[\-/ DIR] [\-n NICE] [\-m BYTES] [\-d BYTES] [\-o N]
	[\-p N] [\-f BYTES] [\-c BYTES] PROG ARGS
.Sp
Change the process state, run PROG
.Sp
.Vb 10
\&        \-u USER[:GRP]   Set uid and gid
\&        \-U USER[:GRP]   Set $UID and $GID in environment
\&        \-e DIR          Set environment variables as specified by files
\&                        in DIR: file=1st_line_of_file
\&        \-/ DIR          Chroot to DIR
\&        \-n NICE         Add NICE to nice value
\&        \-m BYTES        Same as \-d BYTES \-s BYTES \-l BYTES
\&        \-d BYTES        Limit data segment
\&        \-o N            Limit number of open files per process
\&        \-p N            Limit number of processes per uid
\&        \-f BYTES        Limit output file sizes
\&        \-c BYTES        Limit core file size
\&        \-v              Verbose
\&        \-P              Create new process group
\&        \-0              Close stdin
\&        \-1              Close stdout
\&        \-2              Close stderr
.Ve
.IP \fBchroot\fR 4
.IX Item "chroot"
chroot NEWROOT [PROG ARGS]
.Sp
Run PROG with root directory set to NEWROOT
.IP \fBchrt\fR 4
.IX Item "chrt"
chrt \-m | \-p [PRIO] PID | [\-rfobi] PRIO PROG ARGS
.Sp
Change scheduling priority and class for a process
.Sp
.Vb 7
\&        \-m      Show min/max priorities
\&        \-p      Operate on PID
\&        \-r      Set SCHED_RR class
\&        \-f      Set SCHED_FIFO class
\&        \-o      Set SCHED_OTHER class
\&        \-b      Set SCHED_BATCH class
\&        \-i      Set SCHED_IDLE class
.Ve
.IP \fBchvt\fR 4
.IX Item "chvt"
chvt N
.Sp
Change the foreground virtual terminal to /dev/ttyN
.IP \fBcksum\fR 4
.IX Item "cksum"
cksum FILE...
.Sp
Calculate CRC32 checksum of FILEs
.IP \fBclear\fR 4
.IX Item "clear"
clear
.Sp
Clear screen
.IP \fBcmp\fR 4
.IX Item "cmp"
cmp [\-ls] [\-n NUM] FILE1 [FILE2 [SKIP1 [SKIP2]]]
.Sp
Compare FILE1 with FILE2 (or stdin)
.Sp
.Vb 4
\&        \-l      Write the byte numbers (decimal) and values (octal)
\&                for all differing bytes
\&        \-s      Quiet
\&        \-n NUM  Compare at most NUM bytes
.Ve
.IP \fBcomm\fR 4
.IX Item "comm"
comm [\-123] FILE1 FILE2
.Sp
Compare FILE1 with FILE2
.Sp
.Vb 3
\&        \-1      Suppress lines unique to FILE1
\&        \-2      Suppress lines unique to FILE2
\&        \-3      Suppress lines common to both files
.Ve
.IP \fBconspy\fR 4
.IX Item "conspy"
conspy [\-vcsndfFQ] [\-x COL] [\-y LINE] [CONSOLE_NO]
.Sp
A text-mode VNC like program for Linux virtual consoles.
To exit, quickly press ESC 3 times.
.Sp
.Vb 10
\&        \-v      Don\*(Aqt send keystrokes to the console
\&        \-c      Create missing /dev/{tty,vcsa}N
\&        \-s      Open a SHELL session
\&        \-n      Black & white
\&        \-d      Dump console to stdout
\&        \-f      Follow cursor
\&        \-F      Assume console is on a framebuffer device
\&        \-Q      Disable exit on ESC\-ESC\-ESC
\&        \-x COL  Starting column
\&        \-y LINE Starting line
.Ve
.IP \fBcp\fR 4
.IX Item "cp"
cp [\-arPLHpfinlsTu] SOURCE DEST
or: cp [\-arPLHpfinlsu] SOURCE... { \-t DIRECTORY | DIRECTORY }
.Sp
Copy SOURCEs to DEST
.Sp
.Vb 10
\&        \-a      Same as \-dpR
\&        \-R,\-r   Recurse
\&        \-d,\-P   Preserve symlinks (default if \-R)
\&        \-L      Follow all symlinks
\&        \-H      Follow symlinks on command line
\&        \-p      Preserve file attributes if possible
\&        \-f      Overwrite
\&        \-i      Prompt before overwrite
\&        \-n      Don\*(Aqt overwrite
\&        \-l,\-s   Create (sym)links
\&        \-T      Refuse to copy if DEST is a directory
\&        \-t DIR  Copy all SOURCEs into DIR
\&        \-u      Copy only newer files
.Ve
.IP \fBcpio\fR 4
.IX Item "cpio"
cpio [\-dmvu] [\-F FILE] [\-R USER[:GRP]] [\-H newc] [\-tio] [\-p DIR] [EXTR_FILE]...
.Sp
Extract (\-i) or list (\-t) files from a cpio archive, or
take file list from stdin and create an archive (\-o) or copy files (\-p)

sources/core/busybox.py
from pathlib import Path
from utils.load import load_config
from utils.download import download_file, extract_tarball
from utils.execute import run_command_live, run_command
import os
import multiprocessing

DEFAULT_PATCH = {
    "CONFIG_TC": "n",           # TC deaktivieren
    "CONFIG_STATIC": "y",       # Optional: statisches Binary
}


def set_config_option(cfg_file: Path, key: str, value: str):
    """Setzt oder ersetzt eine Option in der .config"""
    lines = cfg_file.read_text().splitlines()
    for i, line in enumerate(lines):
        if line.startswith(f"{key}="):
            lines[i] = f"{key}={value}"
            break
    else:
        lines.append(f"{key}={value}")
    cfg_file.write_text("\n".join(lines) + "\n")



def parse_patch_list(patch_list):
    """Wandelt eine Liste von Strings wie 'KEY=VALUE' in ein Dict um, ignoriert Kommentare"""
    patch_dict = {}
    for line in patch_list:
        line = line.strip()
        if not line or line.startswith("#"):
            continue
        if "=" in line:
            key, val = line.split("=", 1)
            patch_dict[key.strip()] = val.strip()
    return patch_dict



def patch_config(busybox_src_dir: Path, patch_options: dict):
    """Patched die .config Datei mit den gegebenen Optionen"""
    cfg_file = busybox_src_dir / ".config"
    if not cfg_file.exists():
        raise FileNotFoundError(f".config nicht gefunden in {busybox_src_dir}")

    for key, val in patch_options.items():
        set_config_option(cfg_file, key, val)
    print(f"Console > .config gepatcht: {list(patch_options.keys())}")






def build_busybox(args, work_dir: Path, downloads_dir: Path, rootfs_dir: Path):
    """Loads, Extracts, Configures, Compilies and Installs Busybox into target FS"""
    
    # Load Config
    config = load_config(Path("configs") / args.config)
    version = config["version"]
    url = config["url"]
    
    src_dir_template = config["src_dir"]    
    busybox_src_dir = Path(src_dir_template.format(version=version))
    
    cross_compile = config.get("cross_compile", {})
    
    extra_cfg = config.get("extra_config", {})  
    config_patches = config.get("config_patch", [])
    config_patch_dict = parse_patch_list(config_patches)

    print("Console > BusyBox Source Dir:", busybox_src_dir)
    
    
    
    
    # Adjust Architecture
    if args.arch:
        cross_compile["arch"] = args.arch
        if args.arch == "x86_64":
            cross_compile["compiler_prefix"] = ""
        elif args.arch == "arm64":
            cross_compile["compiler_prefix"] = "aarch64-linux-gnu-"

    # Paths
    downloads_dir.mkdir(parents=True, exist_ok=True)
    rootfs_dir.mkdir(parents=True, exist_ok=True)

    # Download & Extract
    print(f"Console > Lade BusyBox {version} herunter...")
    tarball = download_file(url, downloads_dir)

    print(f"Console > Entpacke BusyBox {version}...")
    extracted_dir = extract_tarball(tarball, work_dir)

    # Get Source Location
    subdirs = [d for d in extracted_dir.iterdir() if d.is_dir()]
    src_dir = subdirs[0] if len(subdirs) == 1 else extracted_dir
    
    print(f"Console > BusyBox Quellverzeichnis: {busybox_src_dir}")

    # Enviroment Variables for Cross-Compile
    env = os.environ.copy()
    arch = cross_compile.get("arch", "arm64")
    env["ARCH"] = arch
    if arch != "x86_64":
        env["CROSS_COMPILE"] = cross_compile.get("compiler_prefix", "")
    env["CFLAGS"] = cross_compile.get("cflags", "")
    env["LDFLAGS"] = cross_compile.get("ldflags", "")



    # 1️⃣ defconfig created
    run_command_live(["make", "defconfig"], cwd=busybox_src_dir, env=env, desc="BusyBox defconfig erstellen")

    # 2️⃣ .config patch (TC deactivated + optional extra_cfg)
    print("Console > Patching Busybox's: -> .config - file !.. .. . \n With:")
    patch_dict = parse_patch_list(config["config_patch"])
    print(patch_dict)
    print(extra_cfg)
    patch_config(busybox_src_dir, {**DEFAULT_PATCH, **config_patch_dict, **extra_cfg})

    
    # 3️⃣ oldconfig non-interaktiv
    run_command_live(
        ["make", "oldconfig", "KCONFIG_ALLCONFIG=/dev/null"],
        cwd=busybox_src_dir,
        env=env,
        desc="BusyBox oldconfig (non-interaktiv)"
    )


    # 4️⃣ Kompilieren mit allen Cores
    print("Detecting available CPU-Cores for compiling source-code ... .. .")
    num_cores = multiprocessing.cpu_count()
    print(f"Detected: {num_cores}")
    
    print(f"Console > Compiling BusyBox with {num_cores} Cores...")
    run_command_live(["make", f"-j{num_cores}"], cwd=busybox_src_dir, env=env, desc="BusyBox kompilieren")

    # 5️⃣ Installation ins RootFS
    run_command_live(["make", f"CONFIG_PREFIX={rootfs_dir}", "install"], cwd=busybox_src_dir, env=env, desc="BusyBox installieren")

    print(f"✅ BusyBox {version} successfully installed in {rootfs_dir}")
    
    
    
